hibernate-core/src/main/java/org/hibernate/collection/internal/PersistentBag.java

```diff
@@ -131,21 +134,46 @@ public void beforeInitialize(CollectionPersister persister, int anticipatedSize)
 	}
 
 	@Override
+	@SuppressWarnings( "unchecked" )
 	public boolean equalsSnapshot(CollectionPersister persister) throws HibernateException {
 		final Type elementType = persister.getElementType();
-		final List sn = (List) getSnapshot();
+		final List<Object> sn = (List<Object>) getSnapshot();
 		if ( sn.size() != bag.size() ) {
 			return false;
 		}
-		for ( Object elt : bag ) {
-			final boolean unequal = countOccurrences( elt, bag, elementType ) != countOccurrences( elt, sn, elementType );
-			if ( unequal ) {
+
+		// HHH-11032 - Group objects by Type.getHashCode() to reduce the complexity of the search
+		final Map<Integer, List<Object>> hashToInstancesBag = groupByEqualityHash( bag, elementType );
+		final Map<Integer, List<Object>> HashToInstancesSn = groupByEqualityHash( sn, elementType );
+		if ( hashToInstancesBag.size() != HashToInstancesSn.size() ) {
+			return false;
+		}
+
+		for ( Map.Entry<Integer, List<Object>> hashToInstancesBagEntry : hashToInstancesBag.entrySet() ) {
+			final int hash = hashToInstancesBagEntry.getKey();
+			final List<Object> instancesBag = hashToInstancesBagEntry.getValue();
+			final List<Object> instancesSn = HashToInstancesSn.getOrDefault( hash, Collections.emptyList() );
+			if ( instancesBag.size() != instancesSn.size() ) {
 				return false;
 			}
+			for ( Object instance : instancesBag ) {
+				if ( countOccurrences( instance, instancesBag, elementType ) != countOccurrences( instance, instancesSn, elementType ) ) {
```

Comment:
there's another set of possible optimisations here:

1# don't count on both lists, but rather count on one and then start iterating on the other and see if you can match the same figure. You will be able to quit the loop early, if you reach the figure before ending the loop.

2 # A more tricky to code alternative one would be to not count _neither_ list, and increment the iterator on turns attempting to match the largest figure found so far; essentially you'd also quite early,  but always based on the smallest set of matches - so safe a couple more iterations, but also possibly cheaper to verify each condition.

I'd go with the first proposal as the second is probably too funky: far less readable and I'm not sure it's worth the difference.

Reply:
For 1#, I assume you meant "You will be able to quit the loop early, if you *exceed* the figure before ending the loop."

That's a good idea. I'll make that change.

Reply:
exactly! :+1: